File: Commons/Injector/src/main/java/com/greentree/commons/injector/FieldDependency.java
package com.greentree.commons.injector;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.Objects;
import java.util.Optional;

public record FieldDependency(Field field) implements Dependency {
	
	public FieldDependency {
		Objects.requireNonNull(field);
		var mod = field.getModifiers();
		if(Modifier.isStatic(mod))
			throw new IllegalArgumentException("static field");
		if(Modifier.isFinal(mod))
			throw new IllegalArgumentException("final field");
	}
	
	@Override
	public String toString() {
		return "FieldDependency [" + field + "]";
	}
	
	@Override
	public void set(Object host, InjectionContainer container) {
		try {
			var optional = valueOptional(container);
			if(optional.isEmpty())
				throw new IllegalArgumentException("not found value to " + field + " to host:" + host);
			var access = field.canAccess(host);
			try {
				field.setAccessible(true);
				field.set(host, optional.get());
			}finally {
				field.setAccessible(access);
			}
		}catch(IllegalArgumentException | IllegalAccessException e) {
			throw new RuntimeException(e);
		}
	}
	
	private Optional<?> valueOptional(InjectionContainer container) {
		var value = container.get(field.getName(), field.getType());
		if(value.isPresent())
			return value;
		return container.get(field.getType());
	}
	
}

File: Commons/Injector/src/main/java/com/greentree/commons/injector/FieldDependency.kt
package com.greentree.commons.injector

import com.greentree.commons.util.UnsafeUtil
import java.lang.reflect.Field
import java.lang.reflect.Modifier
import java.util.*

abstract class FieldDependency(val field: Field) : Dependency {

	override fun toString(): String {
		return "FieldDependency [$field]"
	}

	val offset = UnsafeUtil.getUnsafeInstance().objectFieldOffset(field)

	private fun valueOptional(container: InjectionContainer): Optional<*> {
		return container[field.name, field.type]
	}

	override fun set(host: Any, container: InjectionContainer) {
		try {
			val optional = valueOptional(container)
			require(!optional.isEmpty) { "not found value to $field to host:$host" }
			setValue(host, optional.get())
		} catch(e: IllegalArgumentException) {
			throw RuntimeException(e)
		} catch(e: IllegalAccessException) {
			throw RuntimeException(e)
		}
	}

	abstract fun setValue(host: Any, value: Any)

	init {
		Objects.requireNonNull(field)
		val mod = field.modifiers
		require(!Modifier.isStatic(mod)) { "static field" }
		require(!Modifier.isFinal(mod)) { "final field" }
	}
}
